drivers/serial/uart_efinix_sapphire.c

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2023 Efinix Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT efinix_sapphire_uart0
+
+#include "uart_efinix_sapphire_priv.h"
+#include <zephyr/device.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/types.h>
+
+#include <soc.h>
+
+#define UART0_BASE_ADDR DT_INST_REG_ADDR(0)
+
+#define UART0_DATA_REG_ADDR   UART0_BASE_ADDR + BSP_UART_DATA
+#define UART0_STATUS_REG_ADDR UART0_BASE_ADDR + BSP_UART_STATUS
+#define UART0_CLOCK_REG_ADDR  UART0_BASE_ADDR + BSP_UART_CLOCK_DIVIDER
+#define UART0_FRAME_REG_ADDR  UART0_BASE_ADDR + BSP_UART_FRAME_CONFIG
+
+#define UART_IRQ DT_INST_IRQN(0)
+
+struct uart_efinix_sapphire_config {
+	uint32_t baudrate;
+};
+
+struct uart_efinix_sapphire_data {
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t callback;
+	void *cb_data;
+#endif
+};
+
+static void uart_efinix_sapphire_poll_out(const struct device *dev, unsigned char c)
+{
+	/* uart_writeAvailability */
+	while ((sapphire_msb_read16(UART0_STATUS_REG_ADDR) & 0xFF) == 0) {
+	}
+
+	sapphire_write8(UART0_DATA_REG_ADDR, c);
+}
+
+static int uart_efinix_sapphire_poll_in(const struct device *dev, unsigned char *c)
+{
+
+	if ((sapphire_read32(UART0_STATUS_REG_ADDR) & 0xFF000000) != 0) {
+		*c = (unsigned char)sapphire_read32(UART0_DATA_REG_ADDR);
+		return 0;
+	}
+
+	return -1;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+static void uart_efinix_sapphire_irq_tx_enable(const struct device *dev)
+{
+	sapphire_write32(UART0_STATUS_REG_ADDR, sapphire_read32(UART0_STATUS_REG_ADDR) | 0x1);
+}
+
+static void uart_efinix_sapphire_irq_tx_disable(const struct device *dev)
+{
+	sapphire_write32(UART0_STATUS_REG_ADDR, sapphire_read32(UART0_STATUS_REG_ADDR) & ~0x1);
+}
+
+static void uart_efinix_sapphire_irq_rx_enable(const struct device *dev)
+{
+	sapphire_write32(UART0_STATUS_REG_ADDR, sapphire_read32(UART0_STATUS_REG_ADDR) | 0x2);
+}
+
+static void uart_efinix_sapphire_irq_rx_disable(const struct device *dev)
+{
+	sapphire_write32(UART0_STATUS_REG_ADDR, sapphire_read32(UART0_STATUS_REG_ADDR) & ~0x2);
+}
+
+static int uart_efinix_sapphire_irq_tx_ready(const struct device *dev)
+{
+	if (sapphire_read32(UART0_STATUS_REG_ADDR) & 0x100) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static int uart_efinix_sapphire_irq_rx_ready(const struct device *dev)
+{
+	/* Check bit 9 of uart status return true if set else false*/
+	if (sapphire_read32(UART0_STATUS_REG_ADDR) & 0x200) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static int uart_efinix_sapphire_fifo_fill(const struct device *dev, const uint8_t *tx_data, int len)
```

Comment:
Why are we ignoring the `len` argument? We should be writing here as many bytes from the `tx_data` buffer as we can fit in the FIFO.